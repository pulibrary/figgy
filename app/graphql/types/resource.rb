# frozen_string_literal: true
module Types::Resource
  include Types::BaseInterface
  description "A resource in the system."
  orphan_types Types::ScannedResourceType, Types::FileSetType, Types::ScannedMapType

  field :id, String, null: true
  field :label, String, null: true
  field :viewing_hint, String, null: true
  field :url, String, null: true
  field :members, [Types::Resource], null: true
  field :source_metadata_identifier, String, null: true
  field :thumbnail, Types::Thumbnail, null: true

  definition_methods do
    def resolve_type(object, _context)
      "Types::#{object.class}Type".constantize
    end
  end

  def members
    @members ||= Wayfinder.for(object).members
  end

  def manifest_url
    helper.polymorphic_url([:manifest, object])
  end

  def url
    @url ||= helper.show_url(object)
  end

  # We need to centralize logic for navigating a MVW's members to find a
  # thumbnail file set. This is a hack to use the helper's logic for doing that.
  # refactor ticketed as https://github.com/pulibrary/figgy/issues/1708
  def helper
    @helper ||= ManifestBuilder::ManifestHelper.new.tap do |helper|
      helper.singleton_class.include(ThumbnailHelper)
      helper.define_singleton_method(:image_tag) do |url, _opts|
        url
      end
      helper.define_singleton_method(:image_path) do |url|
        url
      end
    end
  end

  def thumbnail
    return if object.try(:thumbnail_id).blank? || thumbnail_resource.blank?
    {
      id: thumbnail_resource.id.to_s,
      thumbnail_url: helper.figgy_thumbnail_path(thumbnail_resource),
      iiif_service_url: helper.figgy_thumbnail_path(thumbnail_resource).gsub("/full/!200,150/0/default.jpg", "")
    }
  end

  def thumbnail_resource
    @thumbnail_resource ||= query_service.find_by(id: object.try(:thumbnail_id).first)
  rescue Valkyrie::Persistence::ObjectNotFoundError
    nil
  end

  def query_service
    Valkyrie::MetadataAdapter.find(:indexing_persister).query_service
  end
end
