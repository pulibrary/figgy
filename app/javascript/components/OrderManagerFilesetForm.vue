<template>
  <div>
    <heading level="h2">
      Edit <small>the selected item</small>
    </heading>
    <form
      id="app"
      novalidate="true"
    >
      <input-text
        id="itemLabel"
        v-model="singleForm.caption"
        label="Label"
        placeholder="e.g., example.tif"
        @input="updateSingle()"
      />

      <input-select
        v-if="!isMultiVolume"
        id="pageType"
        label="Page Type"
        :value="singleForm.viewingHint"
        :options="viewHintOpts"
        @change="updateViewHint($event)"
      />

      <input-checkbox
        v-if="!isMultiVolume"
        id="startCanvasCheckbox"
        :options="startCanvasOpts"
        @change="updateStartCanvas($event)"
      />
      <input-checkbox
        id="thumbnailCheckbox"
        :options="thumbnailOpts"
        @change="updateThumbnail($event)"
      />
    </form>
  </div>
</template>

<script>
import { mapState } from 'vuex'
/**
 * This is the Fileset Form for the Order Manager in Figgy.
 * The only reason you can currently see this pattern is because hiding it causes an OpenSeadragon bug in the Order Manager.
 * Nothing to see here...
 */
export default {
  name: 'FilesetForm',
  status: 'ready',
  release: '1.0.0',
  type: 'Pattern',
  metaInfo: {
    title: 'Fileset Form',
    htmlAttrs: {
      lang: 'en'
    }
  },
  props: {
    /**
     * The html element name used for the component.
     */
    type: {
      type: String,
      default: 'div'
    }
  },
  computed: {
    ...mapState({
      resource: state => state.ordermanager.resource,
      gallery: state => state.gallery
    }),
    memberViewHint: function () {
      let id = this.gallery.selected[0].id
      let selectedMember = this.resource.members.find(member => member.id === id)
      return selectedMember.viewingHint
    },
    startCanvas: function () {
      return this.resource.startCanvas
    },
    thumbnail: function () {
      return this.resource.thumbnail
    },
    isMultiVolume () {
      return this.$store.getters.isMultiVolume
    },
    isStartCanvas: function () {
      let id = this.gallery.selected[0].id
      return this.resource.startCanvas === id
    },
    isThumbnail: function () {
      let id = this.gallery.selected[0].id
      return this.resource.thumbnail === id
    },
    startCanvasOpts: function () {
      return [
        {
          name: 'isStartCanvas',
          value: 'Set as Start Page',
          id: 'isStartCanvas',
          checked: this.isStartCanvas
        }
      ]
    },
    thumbnailOpts: function () {
      return [
        {
          name: 'isThumbnail',
          value: 'Set as Resource Thumbnail',
          id: 'isThumbnail',
          checked: this.isThumbnail
        }
      ]
    },
    viewHintOpts: function () {
      return [
        { label: 'Single Page (Default)', value: 'single', selected: this.isViewHint('single') },
        { label: 'Non-paged', value: 'non-paged', selected: this.isViewHint('non-paged') },
        { label: 'Facing Pages', value: 'facing', selected: this.isViewHint('facing') }
      ]
    },
    singleForm () {
      return {
        caption: this.gallery.selected[0].caption,
        id: this.gallery.selected[0].id,
        mediaUrl: this.gallery.selected[0].mediaUrl,
        service: this.gallery.selected[0].service,
        title: this.gallery.selected[0].title,
        viewingHint: this.gallery.selected[0].viewingHint
      }
    }
  },
  methods: {
    isViewHint (hint) {
      return this.singleForm.viewingHint === hint
    },
    updateStartCanvas (checked) {
      let startCanvas = null
      if (checked) {
        startCanvas = this.gallery.selected[0].id
      }
      this.$store.dispatch('updateStartCanvas', startCanvas)
    },
    updateThumbnail (checked) {
      let thumbnail = null
      if (checked) {
        thumbnail = this.gallery.selected[0].id
      } else {
        thumbnail = null
      }
      this.$store.dispatch('updateThumbnail', thumbnail)
    },
    updateViewHint (event) {
      let viewHint = this.gallery.selected[0].viewingHint
      if (event) {
        viewHint = event
      }
      this.singleForm.viewingHint = viewHint
      this.updateSingle()
    },
    updateSingle () {
      var changeList = this.gallery.changeList
      var items = this.gallery.items
      var index = this.gallery.items
        .map(function (item) {
          return item.id
        })
        .indexOf(this.gallery.selected[0].id)
      items[index] = this.singleForm

      if (changeList.indexOf(this.gallery.selected[0].id) === -1) {
        changeList.push(this.gallery.selected[0].id)
      }

      this.$store.dispatch('updateChanges', changeList)
      this.$store.dispatch('updateItems', items)
    }
  }
}
</script>

<style lang="scss" scoped>
small {
  font-size: 1rem;
  font-weight: 400;
}
.lux-vertical {
  display: block;
}
</style>
