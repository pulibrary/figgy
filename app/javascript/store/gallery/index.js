export const galleryState = {
  items: [],
  selected: [],
  cut: [],
  changeList: [],
  ogItems: [],
}

export const galleryMutations = {
  CUT(state, itemArray) {
    state.cut = [...itemArray]
  },
  PASTE(state, itemArray) {
    state.items = [...itemArray]
  },
  SELECT(state, itemArray) {
    state.selected = [...itemArray]
  },
  SET_GALLERY(state, items) {
    state.items = items
    state.ogItems = items
  },
  SORT_ITEMS(state, value) {
    state.items = [...value]
  },
  UPDATE_CHANGES(state, changeList) {
    state.changeList = [...changeList]
  },
  UPDATE_ITEMS(state, items) {
    state.items = [...items]
  },
}

export const galleryModule = {
  state: galleryState,
  mutations: galleryMutations,
}
