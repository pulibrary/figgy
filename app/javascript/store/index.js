import Vue from 'vue/dist/vue.esm'
import Vuex from 'vuex'
import actions from './vuex/actions'
import {modules} from 'lux-design-system'
Vue.use(Vuex)

const store = new Vuex.Store({
  actions,
  modules: {
    ordermanager: modules.resourceModule,
    gallery: modules.galleryModule,
  }
})

export default store
