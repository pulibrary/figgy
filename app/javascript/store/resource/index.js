export const resourceState = {
  resource: {
    id: "",
    resourceClassName: "",
    bibId: "",
    label: "Resource not available.",
    thumbnail: "",
    startCanvas: "",
    isMultiVolume: false,
    viewingHint: null,
    viewingDirection: null,
    members: [],
    loadState: "NOT_LOADED",
    saveState: "NOT_SAVED",
    ogState: {},
  },
}

export const resourceMutations = {
  APPLY_STATE(state) {
    state.gallery.ogItems = state.gallery.items
    state.gallery.changeList = []
    state.resource.ogState = {
      startCanvas: resource.startPage,
      thumbnail: resource.thumbnail,
      viewingHint: resource.viewingHint,
      viewingDirection: resource.viewingDirection,
    }
    state.resource.saveState = "NOT_SAVED"
  },
  CHANGE_RESOURCE_LOAD_STATE(state, loadState) {
    state.resource.loadState = loadState
  },
  SAVED_STATE(state, saveStatus) {
    state.resource.saveState = saveStatus
  },
  SET_RESOURCE(state, resource) {
    state.resource.id = resource.id
    state.resource.resourceClassName = resource.__typename
    state.resource.label = resource.label
    state.resource.startCanvas = resource.startPage
    state.resource.viewingHint = resource.viewingHint
    state.resource.viewingDirection = resource.viewingDirection
    state.resource.thumbnail = resource.thumbnail != null ? resource.thumbnail.id : null
    state.resource.members = resource.members

    const items = resource.members.map(member => ({
      id: member.id,
      viewingHint: member.viewingHint != null ? member.viewingHint : "single",
      caption: member.label, // member.__typename + " : " + member.id,
      service:
        member["thumbnail"] && typeof(member.thumbnail.iiifServiceUrl) != "undefined"
          ? member.thumbnail.iiifServiceUrl
          : Global.figgy.resource.defaultThumbnail,
      mediaUrl:
        member["thumbnail"] && typeof(member.thumbnail.iiifServiceUrl) != "undefined"
          ? member.thumbnail.iiifServiceUrl + "/full/300,/0/default.jpg"
          : Global.figgy.resource.defaultThumbnail,
    }))
    state.gallery.items = items
    state.gallery.ogItems = items
    state.resource.loadState = "LOADED"
    state.resource.ogState = {
      startCanvas: resource.startPage,
      thumbnail: resource.thumbnail != null ? resource.thumbnail.id : null,
      viewingHint: resource.viewingHint,
      viewingDirection: resource.viewingDirection,
    }
  },
  UPDATE_STARTCANVAS(state, startCanvas) {
    state.resource.startCanvas = startCanvas
  },
  UPDATE_THUMBNAIL(state, thumbnail) {
    state.resource.thumbnail = thumbnail
  },
  UPDATE_VIEWDIR(state, viewDir) {
    state.resource.viewingDirection = viewDir
  },
  UPDATE_VIEWHINT(state, viewHint) {
    state.resource.viewingHint = viewHint
  },
}

export const resourceGetters = {
  getMemberCount: state => {
    return state.resource.members.length
  },
  isMultiVolume: state => {
    const volumes = state.resource.members.filter(member => member.__typename === "ScannedResource")
    return volumes.length > 0 ? true : false
  },
  orderChanged: state => {
    let ogOrder = JSON.stringify(state.gallery.ogItems.map(item => item.id))
    let imgOrder = JSON.stringify(state.gallery.items.map(item => item.id))
    return ogOrder !== imgOrder
  },
  stateChanged: (state, getters) => {
    var propsChanged = []
    propsChanged.push(state.resource.ogState.thumbnail !== state.resource.thumbnail)
    propsChanged.push(state.resource.ogState.startCanvas !== state.resource.startCanvas)
    propsChanged.push(state.resource.ogState.viewingHint !== state.resource.viewingHint)
    propsChanged.push(state.resource.ogState.viewingDirection !== state.resource.viewingDirection)
    propsChanged.push(state.gallery.changeList.length > 0)
    propsChanged.push(getters.orderChanged)
    if (propsChanged.indexOf(true) > -1) {
      return true
    } else {
      return false
    }
  },
}
