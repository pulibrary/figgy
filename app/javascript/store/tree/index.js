export const treeState = {
  selected: null,
}

export const treeMutations = {
  SELECT(state, selectedId) {
    state.selected = selectedId
  },
}
