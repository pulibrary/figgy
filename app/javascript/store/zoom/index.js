export const zoomState = {
  zoomed: null,
}

export const zoomMutations = {
  ZOOM(state, obj) {
    state.zoomed = obj
  },
  RESET_ZOOM(state) {
    state.zoomed = null
  },
}

export const zoomGetters = {
  getIsZoomed: state => {
    return state.zoomed
  },
}
