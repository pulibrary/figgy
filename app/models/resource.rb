# frozen_string_literal: true
class Resource < Valkyrie::Resource
  def self.human_readable_type
    default = @_human_readable_type || name.demodulize.titleize
    I18n.translate("models.#{new.model_name.i18n_key}", default: default)
  end

  def self.can_have_manifests?
    false
  end

  def human_readable_type
    self.class.human_readable_type
  end

  def self.model_name
    ::ActiveModel::Name.new(self)
  end

  # Determines if this is an image resource
  # @return [TrueClass, FalseClass]
  def image_resource?
    false
  end

  # Determines if this is a geospatial resource
  # @return [TrueClass, FalseClass]
  def geo_resource?
    false
  end

  # Determines if this is a media resource
  # @return [TrueClass, FalseClass]
  def media_resource?
    false
  end

  # Virtual property used for stashing pre-loaded objects. Populated by a query.
  attr_accessor :loaded
  def loaded
    @loaded ||= {}
  end
end
