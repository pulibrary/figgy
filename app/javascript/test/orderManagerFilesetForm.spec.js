import Vuex from "vuex"
import { createLocalVue, mount, shallowMount } from "@vue/test-utils"
import OrderManagerFilesetForm from "../components/OrderManagerFilesetForm.vue"
import { resourceMutations, resourceGetters } from "../store/resource"
import { modules } from 'lux-design-system'

// create an extended `Vue` constructor
const localVue = createLocalVue()
localVue.use(Vuex)

let wrapper
let getters
let options
let store
let actions
let items = [
  { id: "a", title: "A", viewingHint: "single", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "b", title: "B", caption: "b", mediaUrl: "https://picsum.photos/600/300/?random" },
  { id: "c", title: "C", caption: "c", mediaUrl: "https://picsum.photos/600/300/?random" },
]

describe("OrderManagerFilesetForm.vue", () => {
  beforeEach(() => {
    actions = {
      updateStartCanvas: jest.fn(),
      updateThumbnail: jest.fn(),
      updateItems: jest.fn(),
      updateChanges: jest.fn()
    }

    const gallery = {
      state: {
        items: items,
        selected: [items[0]],
        cut: [],
        changeList: [],
        ogItems: items,
      },
      mutations: modules.galleryModule.mutations,
    }

    const getters = {
      isMultiVolume: () => false
    }

    const resource = {
      state: {
        resource: {
          id: "",
          resourceClassName: "",
          bibId: "",
          label: "Resource not available.",
          thumbnail: "",
          startCanvas: "",
          isMultiVolume: false,
          viewingHint: "single",
          viewingDirection: null,
          members: [],
          loadState: "NOT_LOADED",
          saveState: "NOT_SAVED",
          ogState: {},
        },
      },
      mutations: resourceMutations,
      getters: getters,
      actions: actions,
      modules: {
        gallery: gallery,
      },
    }

    store = new Vuex.Store({
      modules: {
        ordermanager: resource,
        gallery: gallery,
      },
    })

    options = {
      computed: {
        orderChanged: function () {
          return getters.orderChanged
        },
        isDisabled: function () {
          if (getters.stateChanged) {
            return false
          } else {
            return true
          }
        }
      }
    }

    wrapper = mount(OrderManagerFilesetForm, {
      options,
      localVue,
      store,
      stubs: ["heading","input-select", "input-text", "input-checkbox"],
    })
  })

  it('isViewHint should return true if the value is a viewHint', () => {
    expect(wrapper.vm.isViewHint('single')).toBe(true)
  })

  it('updateStartCanvas action gets called when updateStartCanvas method is called', () => {
    wrapper.vm.updateStartCanvas(true)
    expect(actions.updateStartCanvas).toHaveBeenCalled()
  })

  it('updateThumbnail action gets called when updateStartCanvas method is called', () => {
    wrapper.vm.updateThumbnail(true)
    expect(actions.updateThumbnail).toHaveBeenCalled()
  })

  it('updateSingle should call the updateChanges and updateItems actions', () => {
    wrapper.vm.updateSingle()
    expect(actions.updateChanges).toHaveBeenCalled()
    expect(actions.updateItems).toHaveBeenCalled()
  })

  it('pageType and  should not display for MVWs', () => {
    actions = {
      updateStartCanvas: jest.fn(),
      updateThumbnail: jest.fn(),
      updateItems: jest.fn(),
      updateChanges: jest.fn()
    }

    const gallery = {
      state: {
        items: items,
        selected: [items[0]],
        cut: [],
        changeList: [],
        ogItems: items,
      },
      mutations: modules.galleryModule.mutations,
    }

    const getters = {
      isMultiVolume: () => true
    }

    const mvwResource = {
      state: {
        resource: {
          id: "",
          resourceClassName: "",
          bibId: "",
          label: "Resource not available.",
          thumbnail: "",
          startCanvas: "",
          isMultiVolume: true,
          viewingHint: "single",
          viewingDirection: null,
          members: [],
          loadState: "NOT_LOADED",
          saveState: "NOT_SAVED",
          ogState: {},
        },
      },
      mutations: resourceMutations,
      getters: getters,
      actions: actions,
      modules: {
        gallery: gallery,
      },
    }

    store = new Vuex.Store({
      modules: {
        ordermanager: mvwResource,
        gallery: gallery,
      },
    })

    options = {
      computed: {
        isMultiVolume: function() {
          return getters.isMultiVolume
        }
      }
    }

    wrapper = mount(OrderManagerFilesetForm, {
      options,
      localVue,
      store,
      stubs: ["heading","input-select", "input-text", "input-checkbox"],
    })
    expect(wrapper.find('#pageType').exists()).toBe(false)
    expect(wrapper.find('#startCanvasCheckbox').exists()).toBe(false)
    expect(wrapper.find('#thumbnailCheckbox').exists()).toBe(true)
  })

  it("has the expected html structure", () => {
    expect(wrapper.element).toMatchSnapshot()
  })
})
