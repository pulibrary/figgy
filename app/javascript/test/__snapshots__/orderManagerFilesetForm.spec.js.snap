// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderManagerFilesetForm.vue has the expected html structure 1`] = `
<div>
  <heading-stub
    level="h2"
  >
    
    Edit 
    <small>
      the selected item
    </small>
  </heading-stub>
   
  <form
    id="app"
    novalidate="novalidate"
  >
    <input-text-stub
      id="itemLabel"
      label="Label"
      placeholder="e.g., example.tif"
    />
     
    <input-select-stub
      id="pageType"
      label="Page Type"
      options="[object Object],[object Object],[object Object]"
      value="single"
    />
     
    <input-checkbox-stub
      id="startCanvasCheckbox"
      options="[object Object]"
    />
     
    <input-checkbox-stub
      id="thumbnailCheckbox"
      options="[object Object]"
    />
  </form>
</div>
`;
